File: programmers/77486/77486.js
function solution(enroll, referral, seller, amount) {
  // 알고리즘 실행 코드
  const reversedTree = initReversedTree(enroll, referral);
  const profits = initProfits(enroll);

  // 한 판매인에 대해서 계산
  for (let index = 0; index < seller.length; index++) {
    const s = seller[index];
    const a = amount[index];

    calculateReferralsProfit(s, a);
  }

  return [...profits.values()];

  // 해당 판매자에 의한 수익 분배를 계산
  function calculateReferralsProfit(seller, amount) {
    let currentSeller = seller;
    let profit = amount * 100;

    while (currentSeller !== '-') {
      const currentReferralProfit = profits.get(currentSeller);

      // 추천인에게 보낼 금액 10%
      const profitForReferral = Math.floor(profit / 10);

      // 90%를 자신의 수익으로 계산하여 저장
      profits.set(currentSeller, currentReferralProfit + profit - profitForReferral);

      // 추천인에게 보낼 금액이 없으면 while 문을 끝냄
      if (profitForReferral === 0) {
        break;
      }

      // 그렇지 않은 경우, 다음 반복문을 위해 추천인으로 설정
      currentSeller = reversedTree.get(currentSeller);
      profit = profitForReferral;
    }
  }
}

// child가 parent를 가리키는 그래프 (reversedTree)를 초기화
function initReversedTree(enroll, referral) {
  const result = new Map();
  enroll.forEach((e, index) => {
    result.set(e, referral[index]);
  })
  return result;
}

// enroll 순서에 맞게 Profit을 구성할 수 있도록 seller - profit을 0으로 초기화
function initProfits(enroll) {
  const result = new Map();
  enroll.forEach(e => {
    result.set(e, 0);
  })
  return result;
}

File: programmers/77486/77486.java
import java.util.*;

class Solution {
    
    private Map<String, Integer> profits = new LinkedHashMap<>();
    private Map<String, String> reversedTree = new LinkedHashMap<>();
    
    public int[] solution(String[] enroll, String[] referral, String[] seller, int[] amount) {
        
        initMaps(enroll, referral);
        
        for (int i = 0; i < seller.length; i++) {
            String s = seller[i];
            int a = amount[i];
            calculateProfit(s, a * 100);
        }
        
        int[] result = new int[enroll.length];
        
        for (int i = 0; i < enroll.length; i++) {
            result[i] = profits.get(enroll[i]);
        }
        
        return result;
    }
    
    private void initMaps(String[] enroll, String[] referral) {        
        for (int i = 0; i < enroll.length; i++) {
            reversedTree.put(enroll[i], referral[i]);
            profits.put(enroll[i], 0);
        }
    }
    
    private void calculateProfit(String seller, int profit) {
        while (!"-".equals(seller)) {
            Integer currentProfit = profits.get(seller);

            int addedProfit = profit;
            if (profit >= 10) {
                addedProfit -= profit / 10;    
            } 

            profits.put(seller, currentProfit + addedProfit);

            if (profit < 10) {
                break;
            }

            seller = reversedTree.get(seller);
            profit = profit / 10;
        }
    }
}
